[docs] py docstrings: better conversion of C++ types to python

pmolodo opened the pull request:
### Description of Change(s)

Numerous fixes to translation of C++ types to python types

**NOTE:**
This PR is one of several targeting the python docstring generation process.  To see all these PRs in a branch, see [here](https://github.com/PixarAnimationStudios/OpenUSD/compare/dev...pmolodo:USD:pr/python-docstring-tweaks).

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

pmolodo committed:
[docs] py docstrings: better conversion of C++ types to python

jesschimein commented:
Filed as internal issue #USD-8575

dsyu-pixar commented on docs/python/doxygenlib/cdWriterDocstring.py:
```diff
@@ -35,13 +35,82 @@
+
+# we use the reversed sorted order of the module names, because this ensures that longer names (ie, UsdGeom) come
+# before shorter ones (Usd), so the more exact matches are preferred
+PXR_MODULES_OR_JOINED = "|".join(re.escape(m) for m in reversed(pxrModules))
```
@pmolodo in verifying this change, I'm running into an issue where reversed() isn't supported for dictionaries in Python 3.6 & 3.7 (but is supported in 3.8+).  Additionally, it sounds like dictionaries didn't really support strict ordering at all until Python 3.6 (see https://groups.google.com/g/dev-python/c/76kIoAEKlew). 

Would there be a way to make this work for earlier versions of Python 3?

dsyu-pixar replied:
Alternatively, it'd be nice to have some sort of "requires Python 3.8 or better" check or error message. The current error you get on 3.6 is something along the lines of "Error: 'dict' object is not reversible"

pmolodo commented:
rebased to fix merge conflict...

pmolodo committed:
[docs] py docstrings - don't rely on sorted or reversable py dicts (for py < 3.6)

pmolodo commented:
Pushed changes to make it not rely on dicts being ordered or reversable:

https://github.com/PixarAnimationStudios/OpenUSD/pull/2581/commits/f39826a2915b49829d56985692685d25f05a59b2

dsyu-pixar commented:
Hi @pmolodo thanks for the update! Unfortunately, I now get the following issue (when running with Python 3.6 -- this does not occur when testing with Python 3.9):

```
  File "/home/dsyu/trees/dev-usd/docs/python/doxygenlib/cdWriterDocstring.py", line 552, in __convertTypeName
    tokens = [self.__convertTypeNameToken(x) for x in WORDBREAK_RE.split(ret) if x]
ValueError: split() requires a non-empty pattern match.
```

I think this might be due to an re.split() issue in 3.6, possibly fixed in 3.7 (see https://bugs.python.org/issue43222). 

I realize trying to craft things to work in older versions of Python 3 (that have known issues) can turn into an exercise in futility, so I'm fine with making a min requirement of Python 3.8 or 3.9 -- I believe the VFX 2022 reference platform recommends Python 3.9, and we're trying to test/build with that in mind anyway. What do you think? Maybe just adding a python_version() check somewhere early in convertDoxygen.py would be enough to be transparent enough to anyone using older versions?

pmolodo committed:
[docs] fix for python-2.6 bug with zero-with re.split:

https://bugs.python.org/issue43222

pmolodo commented:
I pushed a version which fixes that `re.split` issue on python-3.6.  I tried to test with python-3.6 myself, but ran into another python-3.6 bug related to `locale.getpreferredencoding()` on the docker image we use for builds (that occurs earlier in the build process), so I can't confirm this is the last python-3.6 related issue.

If another one pops up, I say we just state that you need python-3.7+ to run this script...

pmolodo commented:
Also, fyi - I confirmed that this regexp should give the same results once split with this script:

```python
import inspect
import os
import re

WORDBREAK_RE = re.compile(r"\b")
NONWORD_RE = re.compile(r"(\W+)")

THIS_FILE = os.path.abspath(inspect.getsourcefile(lambda: None) or __file__)
THIS_DIR = os.path.dirname(THIS_FILE)

prim_cpp_path = os.path.join(THIS_DIR, "pxr", "usd", "usd", "prim.cpp")

with open(prim_cpp_path, "r", encoding="utf8") as reader:
    test_cpp_text = reader.read()

def tokenize_1(text: str):
    return [x for x in WORDBREAK_RE.split(text) if x]

def tokenize_2(text: str):
    return [x for x in NONWORD_RE.split(text) if x]

split1 = tokenize_1(test_cpp_text)
split2 = tokenize_2(test_cpp_text)

if split1 == split2:
    print("yay!")
else:
    raise Exception("oh noes")
```
